src/motion/space_groups.F

```diff
@@ -0,0 +1,995 @@
+!> \brief routine indentifies the space group and finds rotation matrices.
+!> \param subsys ...
+!> \param geo_section ...
+!> \param gopt_env ...
+!> \param iounit ...
+!> \par History
+!>      01.2020 created [pcazade]
+!> \author Pierre-André Cazade (first version)
+!> \note  rotation matrices innclude translations and translation symmetry:
+!>        it works with supercells as well.
+! **************************************************************************************************
+   SUBROUTINE identify_space_group(subsys, geo_section, gopt_env, iounit)
+
+      TYPE(cp_subsys_type), INTENT(IN), POINTER          :: subsys
+      TYPE(section_vals_type), INTENT(IN), POINTER       :: geo_section
+      TYPE(gopt_f_type), INTENT(IN), POINTER             :: gopt_env
+      INTEGER                                            :: iounit
+
+      CHARACTER(LEN=*), PARAMETER :: routineN = 'identify_space_group', &
+         routineP = moduleN//':'//routineN
+
+      CHARACTER(LEN=default_path_length)                 :: filename
+      INTEGER                                            :: handle, i, k, n_atom, n_core, n_shell, &
+                                                            n_sr_rep, nparticle, plevel, &
+                                                            shell_index
+      INTEGER, ALLOCATABLE, DIMENSION(:)                 :: atype
+      INTEGER, ALLOCATABLE, DIMENSION(:, :)              :: ranges
+      INTEGER, DIMENSION(:), POINTER                     :: tmp
+      REAL(KIND=dp)                                      :: eps_symmetry
+      REAL(KIND=dp), ALLOCATABLE, DIMENSION(:, :)        :: scoord
+      REAL(KIND=dp), DIMENSION(:), POINTER               :: pol
+      TYPE(cell_type), POINTER                           :: cell
+      TYPE(particle_list_type), POINTER                  :: core_particles, particles, &
+                                                            shell_particles
+      TYPE(spgr_type), POINTER                           :: spgr
+
+      CALL timeset(routineN, handle)
+
+      n_sr_rep = 0
+      nparticle = 0
+      n_atom = 0
+      n_core = 0
+      n_shell = 0
+
+      NULLIFY (particles)
+      NULLIFY (core_particles)
+      NULLIFY (shell_particles)
+
+      NULLIFY (cell)
+      cell => subsys%cell
+      CPASSERT(ASSOCIATED(cell))
+
+      CALL cp_subsys_get(subsys, particles=particles, shell_particles=shell_particles, core_particles=core_particles)
+
+      CPASSERT(ASSOCIATED(particles))
+      n_atom = particles%n_els
+      ! Check if we have other kinds of particles in this subsystem
+      IF (ASSOCIATED(shell_particles)) THEN
+         n_shell = shell_particles%n_els
+         CPASSERT(ASSOCIATED(core_particles))
+         n_core = subsys%core_particles%n_els
+         ! The same number of shell and core particles is assumed
+         CPASSERT(n_core == n_shell)
+      ELSE IF (ASSOCIATED(core_particles)) THEN
+         ! This case should not occur at the moment
```

Comment:
Then maybe put a `CPABORT` here just in case.

Reply:
Done.